<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Joining and Aggregating Datasets &#8212; QuartzBio  documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=12dfc556" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=b3ba4146"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Transforming Datasets" href="transforming_datasets.html" />
    <link rel="prev" title="Dataset Versioning" href="dataset_versioning.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="joining-and-aggregating-datasets">
<h1>Joining and Aggregating Datasets<a class="headerlink" href="#joining-and-aggregating-datasets" title="Permalink to this heading">¶</a></h1>
<section id="joining-datasets">
<h2>Joining Datasets<a class="headerlink" href="#joining-datasets" title="Permalink to this heading">¶</a></h2>
<section id="overview">
<h3>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">¶</a></h3>
<p>The EDP Python library enables users to join two datasets by keys (fields) that are related to each other. The provided datasets join functionality is very similar to the well-known left join in SQL.</p>
<p>The following example joins two public datasets, ClinVar and TCGA, on the variant field found in both datasets:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">quartzbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Filter</span><span class="p">,</span> <span class="n">Dataset</span>

<span class="c1">#Write filter to narrow down results to pathogenic PTEN variants</span>
<span class="n">filters</span> <span class="o">=</span> <span class="n">Filter</span><span class="p">(</span><span class="n">clinical_significance__exact</span><span class="o">=</span><span class="s2">&quot;pathogenic&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">Filter</span><span class="p">(</span><span class="n">gene__exact</span><span class="o">=</span><span class="s2">&quot;PTEN&quot;</span><span class="p">)</span>

<span class="c1">#Query first dataset with filter and specify fields of interest</span>
<span class="n">query_A</span> <span class="o">=</span> <span class="n">Dataset</span><span class="o">.</span><span class="n">get_by_full_path</span><span class="p">(</span><span class="s1">&#39;quartzbio:Public:/ClinVar/5.2.0-20210110/Variants-GRCH37&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">filters</span><span class="o">=</span><span class="n">filters</span><span class="p">,</span> <span class="n">fields</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;variant&#39;</span><span class="p">,</span> <span class="s1">&#39;gene&#39;</span><span class="p">])</span>

<span class="c1">#Query second dataset and specify fields of interest</span>
<span class="n">query_B</span> <span class="o">=</span> <span class="n">Dataset</span><span class="o">.</span><span class="n">get_by_full_path</span><span class="p">(</span><span class="s1">&#39;quartzbio:Public:/TCGA/2.0.0-2018-mc3-v0.2.8/SomaticMutations-GRCh37&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">fields</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;existing_variation&#39;</span><span class="p">])</span>

<span class="c1"># A result of datasets joining is a new Query object</span>
<span class="n">join_query</span> <span class="o">=</span> <span class="n">query_A</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">query_B</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s1">&#39;variant&#39;</span><span class="p">,</span> <span class="n">key_b</span><span class="o">=</span><span class="s1">&#39;variant&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">join_query</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

<span class="c1">#Output:</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624289-89624289-CG&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624274-89624274-G&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;rs587782187,CM110154,COSM346197,COSM921056,COSM28916&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624277-C&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624278-89624279-G&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
<span class="o">...</span>
</pre></div>
</div>
<p>Users can add more datasets to the join query using the same key field. There is no limit to the number of additional joins but they may progressively slow down the query.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="c1">#Query additional dataset to be joined</span>
<span class="n">query_C</span> <span class="o">=</span> <span class="n">Dataset</span><span class="o">.</span><span class="n">get_by_full_path</span><span class="p">(</span><span class="s1">&#39;dataset_path&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">fields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_field&quot;</span><span class="p">])</span>

<span class="c1">#Join additional dataset</span>
<span class="n">join_query</span> <span class="o">=</span> <span class="n">join_query</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">query_C</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s1">&#39;variant&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">join_query</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</pre></div>
</div>
<p>Filtering the output of join queries must be done client-side (i.e. within the user’s Python code). For example, to find all the existing variation identifiers for the known pathogenic PTEN variants:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">join_query</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;existing_variation&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

<span class="c1"># Output:</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624274-89624274-G&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;rs587782187,CM110154,COSM346197,COSM921056,COSM28916&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;variant&#39;</span><span class="p">:</span> <span class="s1">&#39;GRCH37-10-89624275-89624275-T&#39;</span><span class="p">,</span> <span class="s1">&#39;existing_variation&#39;</span><span class="p">:</span> <span class="s1">&#39;CM110130,COSM5153&#39;</span><span class="p">}</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
<section id="advanced-example">
<h3>Advanced Example<a class="headerlink" href="#advanced-example" title="Permalink to this heading">¶</a></h3>
<p>The EDP join method enables users to join two datasets whose keys are lists. However, users must apply the explode <a class="reference external" href="https://quartzbio.github.io/quartzbio-python/expression_functions.html">expression function</a> before joining:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">quartzbio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sb</span>

<span class="n">ds1</span> <span class="o">=</span> <span class="n">sb</span><span class="o">.</span><span class="n">Dataset</span><span class="o">.</span><span class="n">retrieve</span><span class="p">(</span><span class="s2">&quot;dataset_id&quot;</span><span class="p">)</span>
<span class="n">ds2</span> <span class="o">=</span> <span class="n">sb</span><span class="o">.</span><span class="n">Dataset</span><span class="o">.</span><span class="n">retrieve</span><span class="p">(</span><span class="s2">&quot;dataset_id&quot;</span><span class="p">)</span>

<span class="c1"># before joining the records, split each record by the values of ensembl_id</span>
<span class="n">query_A</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">fields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">,</span> <span class="s2">&quot;ensembl_id&quot;</span><span class="p">],</span> <span class="n">annotator_params</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;pre_annotation_expression&quot;</span><span class="p">:</span> <span class="s1">&#39;explode(record, fields=[&quot;ensembl_id&quot;])&#39;</span><span class="p">})</span>
<span class="n">query_B</span> <span class="o">=</span> <span class="n">ds2</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">fields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;ensembl_id&quot;</span><span class="p">])</span>

<span class="n">join_query</span> <span class="o">=</span> <span class="n">query_A</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">query_B</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s2">&quot;ensembl_id&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">join_query</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

<span class="c1"># To then put this into a new dataset</span>
<span class="n">target</span> <span class="o">=</span> <span class="n">sb</span><span class="o">.</span><span class="n">Dataset</span><span class="o">.</span><span class="n">get_or_create_by_full_path</span><span class="p">(</span><span class="s1">&#39;~/join_output&#39;</span><span class="p">)</span>
<span class="n">join_query</span><span class="o">.</span><span class="n">migrate</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="join-parameters">
<h3>Join Parameters<a class="headerlink" href="#join-parameters" title="Permalink to this heading">¶</a></h3>
<p>The join method has the following attributes:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Parameter</strong></p></th>
<th class="head"><p><strong>Value</strong></p></th>
<th class="head"><p><strong>Description</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>query_b</p></td>
<td><p>object</p></td>
<td><p>Query object query_B will be joined with the initial query query_A.</p></td>
</tr>
<tr class="row-odd"><td><p>key</p></td>
<td><p>string</p></td>
<td><p>A key from query_A containing a value that makes a relationship with query_B.</p></td>
</tr>
<tr class="row-even"><td><p>key_b</p></td>
<td><p>string</p></td>
<td><p>(Optional, default=None) a key from query_B containing a value from a key from query_A. If it is None a key argument from query_A will be used.</p></td>
</tr>
<tr class="row-odd"><td><p>prefix</p></td>
<td><p>string</p></td>
<td><p>(Optional, default=b_) a prefix that will be added to all filtered fields from query_B. If it is set to None, a random prefix will be used.</p></td>
</tr>
<tr class="row-even"><td><p>always_prefix</p></td>
<td><p>boolean</p></td>
<td><p>(Optional, default=False) an option to add a prefix either always or only when it is necessary e.g. when both joining keys have the same name.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="aggregating-datasets">
<h2>Aggregating Datasets<a class="headerlink" href="#aggregating-datasets" title="Permalink to this heading">¶</a></h2>
<section id="id1">
<h3>Overview<a class="headerlink" href="#id1" title="Permalink to this heading">¶</a></h3>
<p>Both the EDP Python and R client libraries enable users to perform aggregations to build complex summaries of data. Aggregation queries can be run on datasets with the help of facets. Facets can be used to generate aggregated summaries of string (and date) fields as well as numeric fields, and they automatically work on top of <a class="reference external" href="https://quartzbio.github.io/quartzbio-python/filters.html#overview">queries and filters</a>. Facets can also be nested, which provides an incredibly efficient mechanism to summarize binned or rolled-up data (i.e. data summarized by term or by date).</p>
</section>
<section id="string-and-date-aggregations">
<h3>String and Date Aggregations<a class="headerlink" href="#string-and-date-aggregations" title="Permalink to this heading">¶</a></h3>
<p>For string fields (i.e. categorical fields) and date fields, users can utilize facets to find the total number of unique values as well as a list of the most common values that occur in the dataset. When used with a filtered dataset, the results will represent the filtered subset of data.</p>
<p>The following facet types are supported:</p>
<ul class="simple">
<li><p><strong>terms</strong> (default): Returns a list of the top terms and the number of times they occur (in order of this value). The default number of terms returned at once is 10. Users can set a limit up to 1 million (1,000,000) terms returned.</p></li>
<li><p><strong>count</strong>: Returns the number of unique values in the field. For very large datasets, this is an approximate number.</p></li>
</ul>
<p>Facets will not work for text fields that are indexed (and tokenized) for full-text search. Terms facets are also disabled for _id fields.</p>
<p>Examples in Python:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">quartzbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dataset</span><span class="p">,</span> <span class="n">Filter</span>

<span class="n">query</span> <span class="o">=</span> <span class="n">Dataset</span><span class="o">.</span><span class="n">get_by_full_path</span><span class="p">(</span><span class="s1">&#39;quartzbio:Public:/ClinVar/5.2.0-20210110/Variants-GRCH37&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">()</span>

<span class="c1"># Find the most common genes in ClinVar</span>
<span class="n">query</span><span class="o">.</span><span class="n">facets</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">)</span>

<span class="c1"># Retrieve the number of unique genes in ClinVar</span>
<span class="n">query</span><span class="o">.</span><span class="n">facets</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;facet_type&#39;</span><span class="p">:</span> <span class="s1">&#39;count&#39;</span><span class="p">})</span>

<span class="c1"># Filter ClinVar for only variants that relate to drug response.</span>
<span class="c1"># Which are the most common genes now?</span>
<span class="n">f</span> <span class="o">=</span> <span class="n">quartzbio</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="n">clinical_significance</span><span class="o">=</span><span class="s1">&#39;Drug response&#39;</span><span class="p">)</span>
<span class="n">query</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="p">)</span><span class="o">.</span><span class="n">facets</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">)</span>

<span class="c1"># How many genes are in this filtered query?</span>
<span class="n">query</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="p">)</span><span class="o">.</span><span class="n">facets</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;facet_type&#39;</span><span class="p">:</span> <span class="s1">&#39;count&#39;</span><span class="p">})</span>

<span class="c1"># Now, get the top 100 most common values</span>
<span class="n">query</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="p">)</span><span class="o">.</span><span class="n">facets</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;limit&#39;</span><span class="p">:</span> <span class="mi">100</span><span class="p">}</span>
</pre></div>
</div>
</section>
</section>
<section id="numeric-aggregations">
<h2>Numeric Aggregations<a class="headerlink" href="#numeric-aggregations" title="Permalink to this heading">¶</a></h2>
<p>There are various aggregation options are available for numerical fields (such as float/double, integer/long, and date). Instead of returning “common terms”,  numerical facets can calculate summary statistics, histograms, and percentiles. The following facet types are supported:</p>
<ul class="simple">
<li><p><strong>stats</strong>: Default stats return average, count, maximum, minimum, and sum. Extended stats also include standard deviation, standard deviation lower and upper bounds, sum of squares, and variance.</p></li>
<li><p><strong>histogram</strong>: values are binned according to a provided interval. For numerical fields, the default interval is 100. For dates, the default interval is ‘month’. Histogram intervals must be integers, and will therefore not work for fields with values between 0 and 1 (such as allele frequencies).</p></li>
<li><p><strong>percentiles</strong>: calculates estimated percentiles for a field. By default, returns the following percentiles: 1, 5, 25, 50, 75, 95, 99. Percentiles are approximated and have an 1-5% error for very large datasets.</p></li>
</ul>
<p>Examples</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">quartzbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dataset</span>

<span class="n">query</span> <span class="o">=</span> <span class="n">Dataset</span><span class="o">.</span><span class="n">get_by_full_path</span><span class="p">(</span><span class="s1">&#39;quartzbio:Public:/ClinVar/5.2.0-20210110/Variants-GRCH37&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">()</span>

<span class="c1"># Get extended statistics for a numerical field.</span>
<span class="n">query</span><span class="o">.</span><span class="n">facets</span><span class="p">(</span>
    <span class="o">**</span><span class="p">{</span><span class="s1">&#39;info.ALLELEID&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;facet_type&#39;</span><span class="p">:</span> <span class="s1">&#39;stats&#39;</span><span class="p">,</span> <span class="s1">&#39;extended&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}})</span>

<span class="c1"># Calculate a histogram for genomic position in chromosome 12</span>
<span class="c1"># NOTE: We use **-style notation since &quot;chromosome&quot; and &quot;start&quot; are nested fields</span>
<span class="n">query</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">**</span><span class="p">{</span><span class="s1">&#39;genomic_coordinates.chromosome&#39;</span><span class="p">:</span> <span class="mi">12</span><span class="p">})</span><span class="o">.</span><span class="n">facets</span><span class="p">(</span>
    <span class="o">**</span><span class="p">{</span><span class="s1">&#39;genomic_coordinates.start&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;facet_type&#39;</span><span class="p">:</span> <span class="s1">&#39;histogram&#39;</span><span class="p">,</span>
        <span class="s1">&#39;interval&#39;</span><span class="p">:</span> <span class="mi">1000000</span><span class="p">}})</span>
</pre></div>
</div>
<section id="nested-aggregations">
<h3>Nested Aggregations<a class="headerlink" href="#nested-aggregations" title="Permalink to this heading">¶</a></h3>
<p>Nested aggregations can be used to apply an aggregation query to the result of another aggregation query. For example, given a dataset with patient information, users may want to determine the most common diagnosis age for each cancer type. Users could iterate through each cancer type and run a facets query on the age field, but that would require a large number of expensive API calls. Instead, by using nested aggregations, users can simply construct a facets query within an existing facets query, as in the example below.</p>
<p>At this time, users may only nest term and histogram facets under terms facets. Nesting within histogram facets is not currently supported.</p>
<p>The following example  yields the top ten genes associated with each disease in the public TCGA somatic mutations dataset:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">quartzbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dataset</span>

<span class="c1"># Retrieve the TCGA Somatic Mutations dataset</span>
<span class="n">tcga</span> <span class="o">=</span> <span class="n">Dataset</span><span class="o">.</span><span class="n">get_by_full_path</span><span class="p">(</span><span class="s1">&#39;quartzbio:Public:/TCGA/2.0.0-2018-mc3-v0.2.8/SomaticMutations-GRCh37&#39;</span><span class="p">)</span>

<span class="c1"># Retrieve each disease (terms facets)</span>
<span class="c1"># and the associated gene (hugo_symbol) for each (nested terms facet)</span>
<span class="n">facets</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;disease&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;limit&#39;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
        <span class="s1">&#39;facets&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;hugo_symbol&#39;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s1">&#39;limit&#39;</span><span class="p">:</span> <span class="mi">10</span>
            <span class="p">}</span>
        <span class="p">}</span>
    <span class="p">}</span>
<span class="p">}</span>

<span class="n">results</span> <span class="o">=</span> <span class="n">tcga</span><span class="o">.</span><span class="n">query</span><span class="p">()</span><span class="o">.</span><span class="n">facets</span><span class="p">(</span><span class="o">**</span><span class="n">facets</span><span class="p">)</span>

<span class="c1"># Process the nested results</span>
<span class="k">for</span> <span class="n">disease</span><span class="p">,</span> <span class="n">hugo_symbol</span><span class="p">,</span> <span class="n">subfacets</span> <span class="ow">in</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;disease&#39;</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">disease</span><span class="p">,</span> <span class="n">hugo_symbol</span><span class="p">,</span> <span class="n">subfacets</span><span class="p">[</span><span class="s1">&#39;hugo_symbol&#39;</span><span class="p">])</span>

<span class="c1">#Output:</span>
<span class="c1">#The disease is the first value in each row, followed by the total number of records and then each gene and its count</span>
<span class="n">UCEC</span> <span class="mi">934029</span> <span class="p">[[</span><span class="s1">&#39;TTN&#39;</span><span class="p">,</span> <span class="mi">2961</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;MUC16&#39;</span><span class="p">,</span> <span class="mi">863</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="mi">684</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;DST&#39;</span><span class="p">,</span> <span class="mi">677</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;SYNE1&#39;</span><span class="p">,</span> <span class="mi">642</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;CSMD3&#39;</span><span class="p">,</span> <span class="mi">628</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;RYR2&#39;</span><span class="p">,</span> <span class="mi">613</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;NEB&#39;</span><span class="p">,</span> <span class="mi">569</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;ZFHX4&#39;</span><span class="p">,</span> <span class="mi">551</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;OBSCN&#39;</span><span class="p">,</span> <span class="mi">549</span><span class="p">]]</span>

<span class="n">SKCM</span> <span class="mi">494062</span> <span class="p">[[</span><span class="s1">&#39;TTN&#39;</span><span class="p">,</span> <span class="mi">3059</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;MUC16&#39;</span><span class="p">,</span> <span class="mi">2022</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;DNAH5&#39;</span><span class="p">,</span> <span class="mi">885</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;PCLO&#39;</span><span class="p">,</span> <span class="mi">672</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;LRP1B&#39;</span><span class="p">,</span> <span class="mi">515</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;ANK3&#39;</span><span class="p">,</span> <span class="mi">494</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;GPR98&#39;</span><span class="p">,</span> <span class="mi">482</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;CSMD1&#39;</span><span class="p">,</span> <span class="mi">472</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;DNAH7&#39;</span><span class="p">,</span> <span class="mi">460</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;CSMD2&#39;</span><span class="p">,</span> <span class="mi">453</span><span class="p">]]</span>

<span class="n">COAD</span> <span class="mi">240187</span> <span class="p">[[</span><span class="s1">&#39;TTN&#39;</span><span class="p">,</span> <span class="mi">882</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;APC&#39;</span><span class="p">,</span> <span class="mi">461</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;MUC16&#39;</span><span class="p">,</span> <span class="mi">323</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;SYNE1&#39;</span><span class="p">,</span> <span class="mi">300</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;TP53&#39;</span><span class="p">,</span> <span class="mi">229</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;OBSCN&#39;</span><span class="p">,</span> <span class="mi">220</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;FAT4&#39;</span><span class="p">,</span> <span class="mi">211</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;RYR2&#39;</span><span class="p">,</span> <span class="mi">186</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;NEB&#39;</span><span class="p">,</span> <span class="mi">176</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;KRAS&#39;</span><span class="p">,</span> <span class="mi">171</span><span class="p">]]</span>

<span class="n">LUAD</span> <span class="mi">222076</span> <span class="p">[[</span><span class="s1">&#39;TTN&#39;</span><span class="p">,</span> <span class="mi">935</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;MUC16&#39;</span><span class="p">,</span> <span class="mi">562</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;RYR2&#39;</span><span class="p">,</span> <span class="mi">488</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;CSMD3&#39;</span><span class="p">,</span> <span class="mi">480</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;LRP1B&#39;</span><span class="p">,</span> <span class="mi">392</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;USH2A&#39;</span><span class="p">,</span> <span class="mi">374</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;ZFHX4&#39;</span><span class="p">,</span> <span class="mi">348</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;TP53&#39;</span><span class="p">,</span> <span class="mi">304</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;FLG&#39;</span><span class="p">,</span> <span class="mi">269</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;XIRP2&#39;</span><span class="p">,</span> <span class="mi">268</span><span class="p">]]</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">QuartzBio</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Authentication</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="authentication.html">Authentication</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Vaults and Files</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="exporting_data.html">Exporting Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="importing_data.html">Importing Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="import_parameters.html">Import Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="import_parameters.html#json-jsonl">JSON (JSONL)</a></li>
<li class="toctree-l1"><a class="reference internal" href="vaults_and_objects.html">Vaults and Objects</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Datasets</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="creating_and_migrating_datasets.html">Creating and Migrating Datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="dataset_templates.html">Dataset Templates</a></li>
<li class="toctree-l1"><a class="reference internal" href="dataset_versioning.html">Dataset Versioning</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Joining and Aggregating Datasets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#joining-datasets">Joining Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#aggregating-datasets">Aggregating Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#numeric-aggregations">Numeric Aggregations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="transforming_datasets.html">Transforming Datasets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Expressions</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="expressions.html">Expressions</a></li>
<li class="toctree-l1"><a class="reference internal" href="expression_functions.html">Expression Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Global Search</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="data_discovery.html">Data Discovery</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata_and_global_beacons.html">Metadata and Global Beacons</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Queries and Filters</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="filters.html">Filters</a></li>
<li class="toctree-l1"><a class="reference internal" href="querying_datasets_and_files.html">Querying Datasets and Files</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Changelog</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>


<hr />
<ul>
    
    <li class="toctree-l1"><a href="https://pypi.org/project/quartzbio/">Current Version: v1.2.0</a></li>
    
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="dataset_versioning.html" title="previous chapter">Dataset Versioning</a></li>
      <li>Next: <a href="transforming_datasets.html" title="next chapter">Transforming Datasets</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025 Precision for Medicine, inc..
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.1.2</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/joining_and_aggregating_datasets.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>